Model/Gateway.php

```diff
@@ -0,0 +1,266 @@
+     * @param string $reference the purchase reference
+     * @param string $cvv the card verification value - optional but recommended
+     * @param array|null $fraud_data
+     * @return array
+     * @throws \Exception
+     */
+    public function token_purchase($token, $amount, $reference, $cvv = null, $fraud_data = null)
+    {
+        $customer_ip = $this->get_customer_ip();
+
+        if (function_exists('bcmul')) {
+            $int_amount = intval(bcmul($amount, 100));
+        } else {
+            $multiplied = round($amount * 100);
+            $int_amount = (int)$multiplied;
+        }
+        $payload = [
+            "customer_ip" => $customer_ip,
+            "card_token" => $token,
+            "cvv" => $cvv,
+            "amount" => $int_amount,
+            "reference" => $reference
+            ];
+
+        if (!is_null($fraud_data)) {
+            $payload['fraud'] = $fraud_data;
+        }
+        return $this->do_request("POST", "/purchases", $payload);
+    }
+
+    /**
+     * Performs a refund against the FatZebra gateway
+     * @param string $transaction_id the original transaction ID to be refunded
+     * @param float $amount the amount to be refunded
+     * @param string $reference the refund reference
+     * @return array
+     * @throws \Exception
+     */
+    public function refund($transaction_id, $amount, $reference)
+    {
+        if (is_null($transaction_id) || strlen($transaction_id) === 0) {
+            throw new \InvalidArgumentException("Transaction ID is required");
+        }
+        if (is_null($amount) || strlen($amount) === 0) {
+            throw new \InvalidArgumentException("Amount is required");
+        }
+        if (intval($amount) < 1) {
+            throw new \InvalidArgumentException("Amount is invalid - must be a positive value");
+        }
+        if (is_null($reference) || strlen($reference) === 0) {
+            throw new \InvalidArgumentException("Reference is required");
+        }
+
+        if (function_exists('bcmul')) {
+            $int_amount = intval(bcmul($amount, 100));
+        } else {
+            $multiplied = round($amount * 100);
+            $int_amount = (int)$multiplied;
+        }
+
+        $payload = [
+            "transaction_id" => $transaction_id,
+            "amount" => $int_amount,
+            "reference" => $reference
+            ];
+
+        return $this->do_request("POST", "/refunds", $payload);
+    }
+
+    // TODO: auth/captures
+
+    /************** Private functions ***************/
+
+    /**
+     * Performs the request against the Fat Zebra gateway
+     * @param string $method the request method ("POST" or "GET")
+     * @param string $uri the request URI (e.g. /purchases, /credit_cards etc)
+     * @param array $payload the request payload (if a POST request)
+     * @return array
+     * @throws \Exception
+     */
+    private function do_request($method, $uri, $payload = null)
+    {
+        $curl = curl_init();
```

Comment:
Similar to below, should be instead use something more high-level like Zend HTTP client to do this?

Reply:
Certainly would be ideal, but also want to avoid changing the underlying code _too_ much. Noting also this file existing in the module at all is pretty wrong, but we don't really have the power to properly separate this into a library and include via composer dependency. So, I tried to strike a middle ground and make it Magento-y enough to use it via dependency injection. I'll add a comment to that effect and see if the vendor wants to pursue it further.

Reply:
(I also really like that they've got specific error case handling for timeouts, which is something that's not quite as easy to do when using the more abstract \Magento\Framework\HTTP\Client\Curl)